### PR TITLE
feat: initial import of arcade-machine logic

### PR DESCRIPTION
**Pull request by connor4312**
This is an initial input of the arcade-machine logic into React. The core input/focus logic remains mostly unchanged from the focus and input services in React, except for some changes required by linting.

Mostly for review here is the React bindings, found in `src/components/*` and demonstrated to an extent in the sample app. The state container has also changed; in Angular land, an element could have at most a single arcade-machine directive which controlled all the state, but in React world a single element might have multiple configurations attached to it, so some work was needed for coalescing that.

At @irbrownie's suggestion the arcade-machine bindings are implemented as HOCs. For example,

```tsx
import * as React from 'react';
import { ArcAutoFocus } from 'arcade-machine-react';

class FollowButton extends React.PureComponent {
   // ...
}

// later

const FocusedFollowButton = ArcAutoFocus(FollowButton);

<FocusedFollowButton />
```

I still have to implement and tests some of the fancier features and other parts of the demo. However, basic navigation seems to be working well. I also want to clean up and add tests for some of the existing arcade-machine logic.

### CODE REVIEW

File: test-setup.js [hunk 1]
```diff
@@ -0,0 +1,347 @@
+const { configure } = require('enzyme');
+
+const React = require('react');
```
Comment (connor4312) [inline test-setup.js:3]: note: dropping this in via copypasta was necessary until https://github.com/airbnb/enzyme/issues/1509 gets resolved

File: demo/index.tsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
+import * as React from 'react';
+import { render } from 'react-dom';
+import { ArcRoot, ArcAutoFocus, ArcUp, ArcDown } from '../src';
+
+const AutofocusBox = ArcAutoFocus(
+  class extends React.PureComponent<{ onClick: () => void }> {
```
Comment (irbrownie) [inline demo/index.tsx:6]: this can be just a function that returns JSX.

File: src/components/arc-scope.tsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import { StateContainer } from '../state/state-container';
+
+const isArcScopeSymbol = Symbol();
+
+interface IArcScopeClass<P> extends React.ComponentClass<P> {
+  options: Partial<IArcHandler>;
+  [isArcScopeSymbol]: true;
+}
+
+function isArcScope(type: any): type is IArcScopeClass<any> {
```
Comment (irbrownie) [inline src/components/arc-scope.tsx:16]: Do we have to any?

**Comment by saadataz** [comment]
@connor4312 I am seeing a bunch of linting locally. Can you add travis before merging these PRs?

**Comment by connor4312** [comment]
@saadataz Travis is added and passing in PRs up the stack. This PR doesn't even have tslint.

**Comment by saadataz** [comment]
also we should not checkin package-lock file for dependencies. It will cause duplication issues downstream

**Comment by connor4312** [comment]
It will not. The package-lock file does not affect the published code.

**Pull request merged**